### PR TITLE
Landed slack api update, with enhanced feature set

### PR DESCRIPTION
**Pull request by Awarua-**

### CODE REVIEW

File: core/output/slack.js [hunk 1]
```diff
@@ -92,141 +115,373 @@ renameChannel = function(title, thread) {
...
+			slackTeam.bot_id = body.users[i].id;
+			break;
+		}
 	}
-};
+},
 
+generateChannelsMap = function(data) {
+	var channels = {},
+	i = 0,
```
Comment (mrkno) [inline core/output/slack.js]: You do know you can initialize in loop right?

File: core/shim.js [hunk 2]
```diff
@@ -44,19 +44,26 @@ exports.createPlatformModule = function(platform) {
...
+
 	if (!platform.setTitle) {
 		platform.setTitle = function(title, thread) { // fallback to sending a message
 			platform.sendMessage("If I could set the title of this chat I would set it to \"" + title + "\"", thread);
 		}
 	}
-	
+
+	if (!platform.sendPrivateMessage) {
+		platform.sendPrivateMessage = function(title, thread, senderId) {
```
Comment (mrkno) [inline core/shim.js]: title?

File: README.md [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 <i>It does way more than this now...</i>
 
-Kassy is a modular, easily extensible general purpose chat bot. Small node.js modules can be written for it then placed in the modules directory. They will load on startup (or restart) and become part of what the chat bot provides. Current pre-installed modules are located [here](https://github.com/mrkno/Kassy/tree/master/modules). They include a variety of functionality from getting animated gifs to running arbitrary sandboxed javascript code, voting and giving karma.
+Kassy is a modular, easily extensible general purpose chat bot. Small node.js modules can be written for it then placed in the modules directory. They will load on startup (or restart) and become part of what the chat bot provides. Current pre-installed modules are located [here](https://github.com/mrkno/Kassy/tree/master/modules). They include a variety of functionality from getting animated gifs to running arbitrary sandboxed JavaScript code, voting and giving karma.
```
Comment (mrkno) [inline README.md]: Seriously?

File: config_backup.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (mrkno) [inline config_backup.json]: git rid of

File: core/output/slack.js [hunk 5]
```diff
@@ -1,22 +1,25 @@
-var express = require('express'),
-bodyParser  = require("body-parser"),
-request = require('request'),
-shim = require.once("../shim.js"),
-app = null,
-server = null,
+var request = require('request'),
```
Comment (mrkno) [inline core/output/slack.js]: require.safe(). Even if its slightly broken we want it to be used

File: core/output/slack.js [hunk 6]
```diff
@@ -62,27 +79,27 @@ sendFile = function(type, file, description, thread) {
 			"qs": body
 		},
 		function (error, response, body) {
+			body = JSON.parse(body);
 			if (response.statusCode != 200) {
-				console.log('error: ' + response.statusCode);
+				console.error('error: ' + response.statusCode);
```
Comment (mrkno) [inline core/output/slack.js]: will be printed even when not in debug mode

File: core/output/slack.js [hunk 7]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
-	var body = {"token": slackTeam.slack_token},
-	userMap = [];
+sendTyping = function(thread) {
+	var slackTeams = exports.config.slack_teams,
+	teamInfo = getChannelIdAndTeamId(thread),
+	socket = sockets[teamInfo.team_id];
+
+	if (socket != null) {
+		var body = {
+		'id': slackTeams[teamInfo.team_id].event_id,
```
Comment (mrkno) [inline core/output/slack.js]: indent pls

File: core/output/slack.js [hunk 8]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
-	app.use(bodyParser.json());
-	app.use(bodyParser.urlencoded({ extended: true }));
-
-	platform = shim.createPlatformModule({
-		commandPrefix: exports.config.commandPrefix,
-		sendMessage: sendMessage,
-		sendFile: sendFile,
-		setTitle: renameChannel
-	});
+	s;
```
Comment (mrkno) [inline core/output/slack.js]: s?

File: core/output/slack.js [hunk 9]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
-						if (!userName) {
-							// query slack to see if the user is in the team, they may have been added recently
-							message = lookUpUserAddToTeam(slackTeam, id, matches[j], message, replaceUserIdWithUserName);
-						}
-						else {
-							message = replaceUserIdWithUserName(userName, message, matches[j]);
-						}
-					}
+	if (connectionDetails) {
+		(function (connectionDetails) {
```
Comment (mrkno) [inline core/output/slack.js]: reason for anon function

File: core/output/slack.js [hunk 10]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
+			"method": 'GET',
+			"qs": body
+		},
+		function (error, response, body) {
+			body = JSON.parse(body);
+			if (response.statusCode != 200) {
+					console.debug('error: ' + response.statusCode);
+			}
+			else if (!body.ok) {
+					console.debug('Failed to send message to user with id: ' + teamInfo.channel_id + ', error: ' + body.error);
```
Comment (mrkno) [inline core/output/slack.js]: too much indentation

File: core/output/slack.js [hunk 11]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
 			}
+			else {
+				slackTeams[teamInfo.team_id].channels[body.channel] = {"channel": body.channel, "user": senderId};
+				sendMessage(message, createThreadId(body.channel.id, teamInfo.team_id));
+			}
+		});
+	}
+	else {
+			console.debug('No slack team found!!!');
```
Comment (mrkno) [inline core/output/slack.js]: too much indentation

File: core/output/slack.js [hunk 12]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
+	}
+},
+
+eventReceived = function(event, teamId) {
+	if (shuttingDown) return;
+	switch (event.type) {
+		case 'message':
+		recMessage(event, teamId);
```
Comment (mrkno) [inline core/output/slack.js]: indent pls

File: core/output/slack.js [hunk 13]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
+					}
+					else {
+						message = replaceUserIdWithUserName(userName, message, matches[0]);
+					}
+				}
+				matches = regex.exec(message);
+			}
+		}
+		else {
+				console.debug('No slack team found matching: ' + teamId);
```
Comment (mrkno) [inline core/output/slack.js]: Indentation

File: core/output/slack.js [hunk 14]
```diff
@@ -92,142 +109,385 @@ renameChannel = function(title, thread) {
...
 
-	server.close();
-	server = null;
-	app = null;
-	platform = null;
+	shuttingDown = false;
+	eventReceivedCallback = callback;
+	numSocketsToShutDown = 0;
+
+	console.debug("starting");
```
Comment (mrkno) [inline core/output/slack.js]: starting what?

File: core/output/test.js [hunk 15]
```diff
@@ -10,7 +10,7 @@ exports.start = function(callback) {
 		sendMessage: function(message, thread) {
 			console.log(message);
 		}
-	})
+	});
```
Comment (mrkno) [inline core/output/test.js]: should probably actually be a ','

File: core/platform.js [hunk 16]
```diff
@@ -75,6 +75,7 @@ Platform.prototype.messageRxd = function(api, event) {
 
 Platform.prototype.setModes = function(modes) {
     try {
+		var i = 0;
```
Comment (mrkno) [inline core/platform.js]: why and why so much indent

File: core/startup.js [hunk 17]
```diff
@@ -21,6 +21,7 @@ var selectedModes        = null,
             var Platform = require.once('./platform.js'),
                 platform = new Platform(selectedModes);
             platform.setOnShutdown(checkShutdownCode);
+            console.log("Starting platform");
```
Comment (mrkno) [inline core/startup.js]: no

File: modules/common/kassy.json [hunk 18]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "common",
+  "version": 2.0,
+  "startup": ""
+}
```
Comment (mrkno) [inline modules/common/kassy.json]: common directory should definately NOT have a kassy.json

File: config_backup.json (hunk 4 above)
Comment (Awarua-) [inline config_backup.json]: Done

File: README.md (hunk 3 above)
Comment (Awarua-) [inline README.md]: Meh

File: core/output/slack.js (hunk 5 above)
Comment (Awarua-) [inline core/output/slack.js]: done

File: core/output/slack.js (hunk 6 above)
Comment (Awarua-) [inline core/output/slack.js]: Fixed

File: core/output/slack.js (hunk 7 above)
Comment (Awarua-) [inline core/output/slack.js]: Done

File: core/output/slack.js (hunk 8 above)
Comment (Awarua-) [inline core/output/slack.js]: renamed to socket

File: core/output/slack.js (hunk 9 above)
Comment (Awarua-) [inline core/output/slack.js]: I don't remember, removed

File: core/output/slack.js (hunk 10 above)
Comment (Awarua-) [inline core/output/slack.js]: fixed

File: core/output/slack.js (hunk 11 above)
Comment (Awarua-) [inline core/output/slack.js]: fixed

File: core/output/slack.js (hunk 12 above)
Comment (Awarua-) [inline core/output/slack.js]: fixed

File: core/output/slack.js (hunk 13 above)
Comment (Awarua-) [inline core/output/slack.js]: fixed

File: core/output/slack.js (hunk 14 above)
Comment (Awarua-) [inline core/output/slack.js]: Made more specific

File: core/output/test.js (hunk 15 above)
Comment (Awarua-) [inline core/output/test.js]: fixed

File: core/platform.js (hunk 16 above)
Comment (Awarua-) [inline core/platform.js]: look fine to me?

File: README.md [hunk 19]
```diff
@@ -186,6 +185,14 @@ Arguments:
...
 - `thread`. The thread to send it to. String.
 
+#### `sendPrivateMessage(message, thread, senderId)`
+Sends a message to an individual.
+
+Arguments:
+- `message`. The message to send. String.
+- `thread`. The thread which the associated command which sends the private message. String.
+- `senderId`. The sender to send the message to. String.
+
```
Comment (mrkno) [inline README.md]: returns?

File: core/output/slack.js [hunk 20]
```diff
@@ -92,142 +109,382 @@ renameChannel = function(title, thread) {
...
+		case 'user_change':
+			userChange(event, teamId);
+			break;
+		case 'team_join':
+			userChange(event, teamId);
+			break;
+		case 'team_rename':
+			teamRename(event, teamId);
+			break;
+		default:
```
Comment (mrkno) [inline core/output/slack.js]: blank default?

File: core/platform.js [hunk 21]
```diff
@@ -75,23 +75,26 @@ Platform.prototype.messageRxd = function(api, event) {
 
 Platform.prototype.setModes = function(modes) {
     try {
+        var i = 0;
         if (this.statusFlag !== StatusFlag.NotStarted) {
             throw 'Cannot change mode when it is already started.';
         }
         this.modes = [];
+        console.log(modes);
```
Comment (mrkno) [inline core/platform.js]: why

File: core/platform.js [hunk 22]
```diff
@@ -75,23 +75,26 @@ Platform.prototype.messageRxd = function(api, event) {
...
         for (var i = 0; i < modes.length; i++) {
             var mode = {
                 instance: require.once('./output/' + modes[i]),
                 name: modes[i]
             };
+            console.log(mode);
```
Comment (mrkno) [inline core/platform.js]: why

File: core/platform.js [hunk 23]
```diff
@@ -75,23 +75,26 @@ Platform.prototype.messageRxd = function(api, event) {
 
 Platform.prototype.setModes = function(modes) {
     try {
+        var i = 0;
```
Comment (mrkno) [inline core/platform.js]: You do realize the scope of this will be overridden by the var in the if statement right?

File: core/shim.js [hunk 24]
```diff
@@ -66,7 +67,14 @@ exports.createPlatformModule = function(platform) {
 			platform.commandPrefix = '/';
 		}
 	}
-	
+
+
```
Comment (mrkno) [inline core/shim.js]: too many newlines

File: README.md (hunk 19 above)
Comment (Awarua-) [inline README.md]: removed

File: core/output/slack.js (hunk 20 above)
Comment (Awarua-) [inline core/output/slack.js]: Fixed

File: core/platform.js (hunk 22 above)
Comment (Awarua-) [inline core/platform.js]: Removed

File: core/platform.js (hunk 21 above)
Comment (Awarua-) [inline core/platform.js]: Removed

File: core/platform.js (hunk 23 above)
Comment (Awarua-) [inline core/platform.js]: Removed

File: core/shim.js (hunk 24 above)
Comment (Awarua-) [inline core/shim.js]: Fixed

File: core/output/slack.js [hunk 25]
```diff
@@ -92,142 +109,383 @@ renameChannel = function(title, thread) {
...
+},
+
+recMessage = function(event, teamId) {
+	var slackTeam = exports.config.slack_teams[teamId],
+	userName,
+	id,
+	message = event.text,
+	shimMessage;
+
+
```
Comment (mrkno) [inline core/output/slack.js]: Whitespace

File: core/shim.js [hunk 26]
```diff
@@ -66,8 +67,12 @@ exports.createPlatformModule = function(platform) {
...
 	}
-	
-	return platform;
+
+	if (!platform.sendPrivateMessage) {
+		platform.sendPrivateMessage = function(message, thread, senderId) {
+			platform.sendMessage(message, thread);
+		}
+	}
 };
```
Comment (mrkno) [inline core/shim.js]: you added a prototype but didnt put the return back

**Pull request merged**